### PR TITLE
Changelog for Clippy 1.78 :magic_wand:

### PR DESCRIPTION
**Pull request by xFrednet**
Roses and Violets have colors,
Red and Blue are the two,
I'm getting to the end of my masters,
what a cool goal to pursue

---

### The cat of this release is: *Shadow* submitted by @benwh1:

The cats of this Clippy release

Cats for the next release can be nominated in the comments :D

---

changelog: none

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -6,11 +6,75 @@ document.
...
+[View all 112 merged pull requests](https://github.com/rust-lang/rust-clippy/pulls?q=merged%3A2024-01-26T05%3A46%3A23Z..2024-03-07T16%3A25%3A52Z+base%3Amaster)
+
+### Important Changes
+
+* In a few versions, `cargo-clippy` will now no longer ignore the first argument.
+  Passing in `clippy` as the first argument like cargo does will continue to work.
+  However, arguments like `--fix` will soon issue a warning and will no longer be.
+  `cargo clippy` (without `-`!) is unaffected by this change.
+  [#9461](https://github.com/rust-lang/rust-clippy/pull/9461)
+
```
Comment (xFrednet) [inline CHANGELOG.md:24]: @flip1995 && @matthiaskrgr can you confirm that this part reflects the planned changes from #9461 correctly?

File: CHANGELOG.md [hunk 2]
```diff
@@ -6,11 +6,75 @@ document.
...
+* [`unnecessary_result_map_or_else`]
+  [#12169](https://github.com/rust-lang/rust-clippy/pull/12169)
+* [`to_string_trait_impl`]
+  [#12122](https://github.com/rust-lang/rust-clippy/pull/12122)
+* [`incompatible_msrv`]
+  [#12160](https://github.com/rust-lang/rust-clippy/pull/12160)
+
+### Moves and Deprecations
+
+* Moved [`mixed_attributes_style`] to `style` (Remains warn-by-default)
```
Comment (y21) [inline CHANGELOG.md:56]: Does this need to be in the release notes considering that the lint is being added in the same release and didn't exist before? (As far as stable is concerned this lint started out in the style category and was never in a different category to begin with)

File: CHANGELOG.md [hunk 3]
```diff
@@ -6,11 +6,75 @@ document.
...
+* [`thread_local_initializer_can_be_made_const`]: Now checks the [`msrv`] configuration
+  [#12405](https://github.com/rust-lang/rust-clippy/pull/12405)
+* [`disallowed_macros`]: Code generated by derive macros can no longer allow this lint
+  [#12267](https://github.com/rust-lang/rust-clippy/pull/12267)
+* [`wildcard_imports`]: Add configuration [`allowed-wildcard-imports`] to allow preconfigured wildcards
+  [#11979](https://github.com/rust-lang/rust-clippy/pull/11979)
+
+### ICE Fixes
+
+* [`ptr_as_ptr`]: No longer ICEs on types from other crates
```
Comment (y21) [inline CHANGELOG.md:70]: Nit
```suggestion
* [`ptr_as_ptr`]: No longer ICEs when the cast source is a function call to a local variable
```
Something like that? That ICE is unrelated to types and was more a problem because it was calling `.def_id()` on a local variable

File: CHANGELOG.md (hunk 2 above)
Comment (xFrednet) [reply CHANGELOG.md:56]: Good point, nope, then it should be removed :+1:

File: CHANGELOG.md [hunk 4]
```diff
@@ -6,11 +6,75 @@ document.
...
+* [`to_string_trait_impl`]
+  [#12122](https://github.com/rust-lang/rust-clippy/pull/12122)
+* [`incompatible_msrv`]
+  [#12160](https://github.com/rust-lang/rust-clippy/pull/12160)
+
+### Moves and Deprecations
+
+* Moved [`mixed_attributes_style`] to `style` (Remains warn-by-default)
+  [#12572](https://github.com/rust-lang/rust-clippy/pull/12572)
+
```
Comment (xFrednet) [inline CHANGELOG.md:58]:
```suggestion
```

**Review by dswij — APPROVED** [review]
r=me once we confirm the `arg` changes to be correct

File: CHANGELOG.md (hunk 1 above)
Comment (flip1995) [reply CHANGELOG.md:24]: > will soon issue a warning and will no longer be.

There's something missing at the end of this sentence.

I think what is missing here is, that arguments like `--fix` will only temporarily emit a warning but will change behavior (removing the warning again) 1-3 release cycles later.

Comment (flip1995) [reply CHANGELOG.md:24]: Also, once we merge the PR that actually introduces the warnings or changes the behavior we might want to write a blog post about it.

**Comment by llogiq** [comment]
Maunzer (a.k.a. Tom), here shown testing the `borrowed_box` lint

**Comment by blyxyas** [comment]
image

File: CHANGELOG.md (hunk 1 above)
Comment (xFrednet) [reply CHANGELOG.md:24]: Good catch. Now that I'm rereading this, why do we want to issue a warning for that? It feels weird that correct usage, like `cargo-clippy --fix` should issue a warning. Or was the idea just to issue a warning for unexpected arguments? Meaning that
`cargo-clippy --fix` would be accepted, but `cargo-clippy duck --fix` will issue a warning?

Comment (xFrednet) [reply CHANGELOG.md:24]: Ping @flip1995 @matthiaskrgr, I've now updated this part to say the following

> * These changes will affect users of the `cargo-clippy` binary. Users of the 
>  `cargo clippy` command will be unaffected by the change.
>  
>    In a few versions, `cargo-clippy` will now no longer ignore the first argument.
>    Passing in `clippy` as the first argument like cargo does will continue to work.
>    There will be a transition period where unknown arguments will issue a warning
>    instead of a hard error.

Does that sound good?

Comment (flip1995) [reply CHANGELOG.md]: Almost. But it never will emit a hard error, it will change bahavior after the transition phase. This is also the answer to why we want to emit a warning first:

Currently `cargo-clippy --fix` does not run `--fix` but will do so in the future. So people using `cargo-clippy --fix` might want to remove the `--fix` flag before the behavior changes, so they can keep the current behavior. Otherwise they can ignore the warning. This warning should also be unaffected by `-Dwarnings`, so that it doesn't break CIs.

**Comment by xFrednet** [comment]
@flip1995 decided to leave the `cargo-clippy` argument thing from this changelog in favor of a more coordinated announcement. The rest was already approved, so I'll merge it now.

Thank you for the cat picture :3

@bors r=dswij

**Comment by Centri3** [comment]
Late to the party this time. But uncovered old photos of kittens when I was 10-11 :)

Gracie and her kittens

**Comment by xFrednet** [comment]
At RustNL I got a truly magnificent cat nomination =^.^=

I present you the one and only @jyn514 :hearts: 

image
